### PR TITLE
Refactoring/factory for di

### PR DESCRIPTION
**Pull request by hromovyi-kostiantyn**
another version of DI factory

### CODE REVIEW

File: src/main/java/executor/service/App.java [hunk 1]
```diff
@@ -1,24 +1,18 @@
...
-    private static String getPackageName() {
-        Package currentPackage = App.class.getPackage();
-        return currentPackage.getName();
-    }
+    ObjectFactoryImpl factory = ObjectFactoryImpl.getInstance();
+    StepExecutionClickCss stepExecutionClickCss1 = factory.create(StepExecutionClickCss.class);
+    StepExecutionClickCss stepExecutionClickCss2 = factory.create(StepExecutionClickCss.class);
+    System.out.println(stepExecutionClickCss1.getStepAction());
+    System.out.println(stepExecutionClickCss2.getStepAction());
+    System.out.println(stepExecutionClickCss1.equals(stepExecutionClickCss2));
```
Comment (jurok3x) [inline src/main/java/executor/service/App.java]: better to put this into tests

**Review by jurok3x — APPROVED** [review]

File: src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    private volatile static ObjectFactoryImpl instance = null;
+    private ObjectFactoryImpl() {}
+    public static ObjectFactoryImpl getInstance() {
+        if (instance == null) {
+            synchronized(ObjectFactoryImpl.class) {
+                if (instance == null) {
+                    instance = new ObjectFactoryImpl();
+                }
+            }
+        }
```
Comment (Sokol3783) [inline src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java]: There are no sense to make like this, ENUM initialize on bootstrap, and it always single.

File: src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ObjectFactoryImpl implements ObjectFactory {
+    private static final Singleton INSTANCE = Singleton.INSTANCE;
+
+    private volatile static ObjectFactoryImpl instance = null;
```
Comment (Sokol3783) [inline src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java]: this lines almost the same, chose one of them

File: src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            }
+
+            for (Type type : declaredConstructor.getParameterTypes()) {
+                Field[] declaredFields = o.getClass().getDeclaredFields();
+                for (Field field : declaredFields) {
+                    if (type.equals(field.getType())) {
+                        field.setAccessible(true);
+                        field.set(field, create(type.getClass()));
+                    }
+                }
```
Comment (Sokol3783) [inline src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java]: What will you insert if it's going to be type Integer or String?
I think better first check constructor contain only interfaces, and only then make insert.
Also we have some classes which can't to create in this way, like implementation of ParallelFlowExecutor, and as I remember, one of webdrivers*** classes, should not be singletones.

File: src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java (hunk 2 above)
Comment (hromovyi-kostiantyn) [reply src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java]: So ObjectFactoryImpl doesn't need to be Singleton?

**Review by Sokol3783 — APPROVED** [review]

File: src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ObjectFactoryImpl implements ObjectFactory {
```
Comment (OlexTul) [inline src/main/java/executor/service/DIFactory/ObjectFactoryImpl.java:12]: According to the contract DI:  "Factory should create all necessary services and using it for dependencies in the service facade".
image

DI creates but does not configure classes. Please indicate how to use this.
image

Dee doesn't work with higher level abstractions.
image
image

Di does not work with classes that do not have an interface, or please give specific recommendations for use.
image
image
image

If I am using DI incorrectly, please give specific recommendations for use.

**Review by OlexTul — CHANGES_REQUESTED** [review]

File: src/main/java/executor/service/App.java [hunk 6]
```diff
@@ -1,24 +1,18 @@
...
-        service.execute();
-    }
+  public static void main(String[] args)
+      throws Throwable {
 
-    private static String getPackageName() {
-        Package currentPackage = App.class.getPackage();
-        return currentPackage.getName();
-    }
+    ObjectFactoryImpl factory = ObjectFactoryImpl.getInstance();
```
Comment (OlexTul) [inline src/main/java/executor/service/App.java]: 1. Give recommendations on creating classes and contracts.
2. Give recommendations on injecting fields into classes.

**Review by OlexTul — CHANGES_REQUESTED** [review]

**Pull request closed without merging**